File: System.app/src/features/backtest/utils/metrics.ts
// src/features/backtest/utils/metrics.ts
// Metrics computation utilities for backtest

import type { BacktestDayRow, EquityPoint } from '@/types'

/**
 * Convert UTC seconds to ISO date string (YYYY-MM-DD)
 */
export const isoFromUtcSeconds = (t: number): string => {
  const ms = Number(t) * 1000
  if (!Number.isFinite(ms)) return '1970-01-01'
  try {
    return new Date(ms).toISOString().slice(0, 10)
  } catch {
    return '1970-01-01'
  }
}

/**
 * Basic metrics computed from equity curve and returns
 */
export type BasicMetrics = {
  cagr: number
  maxDrawdown: number
  sharpe: number
  days: number
}

/**
 * Compute basic metrics from equity curve and returns array
 */
export const computeMetrics = (equity: number[], returns: number[]): BasicMetrics => {
  const days = returns.length
  const final = equity.length ? equity[equity.length - 1] : 1
  const cagr = days > 0 && final > 0 ? final ** (252 / days) - 1 : 0
  let peak = -Infinity
  let maxDd = 0
  for (const v of equity) {
    if (v > peak) peak = v
    if (peak > 0) {
      const dd = v / peak - 1
      if (dd < maxDd) maxDd = dd
    }
  }
  const mean = days > 0 ? returns.reduce((a, b) => a + b, 0) / days : 0
  const variance = days > 1 ? returns.reduce((a, b) => a + (b - mean) ** 2, 0) / (days - 1) : 0
  const std = Math.sqrt(Math.max(0, variance))
  const sharpe = std > 0 ? (Math.sqrt(252) * mean) / std : 0
  return { cagr, maxDrawdown: maxDd, sharpe, days }
}

/**
 * Monthly return bucket
 */
export type MonthlyReturn = {
  year: number
  month: number
  value: number
}

/**
 * Compute monthly returns from daily backtest data
 */
export const computeMonthlyReturns = (days: BacktestDayRow[]): MonthlyReturn[] => {
  const buckets = new Map<string, { year: number; month: number; acc: number }>()
  for (const d of days) {
    const dt = d.date
    const year = Number(dt.slice(0, 4))
    const month = Number(dt.slice(5, 7))
    const key = `${year}-${month}`
    const prev = buckets.get(key) || { year, month, acc: 1 }
    prev.acc *= 1 + (Number.isFinite(d.netReturn) ? d.netReturn : 0)
    buckets.set(key, prev)
  }
  return Array.from(buckets.values())
    .map((b) => ({ year: b.year, month: b.month, value: b.acc - 1 }))
    .sort((a, b) => (a.year - b.year) || (a.month - b.month))
}

/**
 * Summary metrics for a backtest
 */
export type BacktestSummary = {
  startDate: string
  endDate: string
  days: number
  years: number
  totalReturn: number
  cagr: number
  vol: number
  maxDrawdown: number
  calmar: number
  sharpe: number
  sortino: number
  treynor: number
  beta: number
  winRate: number
  bestDay: number
  worstDay: number
  avgTurnover: number
  avgHoldings: number
}

/**
 * Compute comprehensive backtest summary metrics
 */
export const computeBacktestSummary = (
  points: EquityPoint[],
  drawdowns: number[],
  days: BacktestDayRow[],
  benchmarkPoints?: EquityPoint[]
): BacktestSummary => {
  const equity = points.map((p) => p.value)
  const returns = days.map((d) => d.netReturn)
  const base = computeMetrics(equity, returns)

  const totalReturn = equity.length ? equity[equity.length - 1] - 1 : 0
  const years = base.days / 252

  const mean = returns.length ? returns.reduce((a, b) => a + b, 0) / returns.length : 0
  const variance =
    returns.length > 1 ? returns.reduce((a, b) => a + (b - mean) ** 2, 0) / (returns.length - 1) : 0
  const dailyStd = Math.sqrt(Math.max(0, variance))
  const vol = dailyStd * Math.sqrt(252)

  // Sortino: uses downside deviation (squared negative returns relative to 0)
  // Calculate downside deviation using all returns, but only penalizing negative ones
  const downsideSquaredSum = returns.reduce((sum, r) => sum + (r < 0 ? r * r : 0), 0)
  const downsideVariance = returns.length > 1 ? downsideSquaredSum / (returns.length - 1) : 0
  const downsideStd = Math.sqrt(Math.max(0, downsideVariance))
  const annualizedDownsideStd = downsideStd * Math.sqrt(252)
  const annualizedMean = mean * 252
  const sortino = annualizedDownsideStd > 0 ? annualizedMean / annualizedDownsideStd : 0

  // Treynor Ratio: (Portfolio Return - Risk-Free Rate) / Beta
  // Beta = Cov(Rp, Rm) / Var(Rm)
  let treynor = 0
  let beta = 0
  if (benchmarkPoints && benchmarkPoints.length > 1 && returns.length > 1) {
    // Calculate benchmark returns from benchmark equity points
    const benchReturns: number[] = []
    for (let i = 1; i < benchmarkPoints.length; i++) {
      const prev = benchmarkPoints[i - 1].value
      const curr = benchmarkPoints[i].value
      if (prev > 0) {
        benchReturns.push(curr / prev - 1)
      }
    }

    // Align returns arrays (use minimum length)
    const minLen = Math.min(returns.length, benchReturns.length)
    if (minLen > 1) {
      const portReturns = returns.slice(0, minLen)
      const mktReturns = benchReturns.slice(0, minLen)

      const portMean = portReturns.reduce((a, b) => a + b, 0) / minLen
      const mktMean = mktReturns.reduce((a, b) => a + b, 0) / minLen

      // Covariance of portfolio and market
      let cov = 0
      let mktVar = 0
      for (let i = 0; i < minLen; i++) {
        cov += (portReturns[i] - portMean) * (mktReturns[i] - mktMean)
        mktVar += (mktReturns[i] - mktMean) ** 2
      }
      cov /= (minLen - 1)
      mktVar /= (minLen - 1)

      // Beta = Cov(Rp, Rm) / Var(Rm)
      beta = mktVar > 0 ? cov / mktVar : 0

      // Treynor = annualized excess return / beta (assuming 0% risk-free rate)
      if (beta !== 0) {
        treynor = annualizedMean / beta
      }
    }
  }

  const winRate = returns.length ? returns.filter((r) => r > 0).length / returns.length : 0
  const bestDay = returns.length ? Math.max(...returns) : 0
  const worstDay = returns.length ? Math.min(...returns) : 0
  const avgTurnover = days.length ? days.reduce((a, d) => a + d.turnover, 0) / days.length : 0
  const avgHoldings = days.length ? days.reduce((a, d) => a + d.holdings.length, 0) / days.length : 0

  const startDate = points.length ? isoFromUtcSeconds(points[0].time) : ''
  const endDate = points.length ? isoFromUtcSeconds(points[points.length - 1].time) : ''
  const maxDrawdown = drawdowns.length ? Math.min(...drawdowns) : 0

  // Calmar: CAGR / abs(maxDrawdown)
  const calmar = maxDrawdown !== 0 ? base.cagr / Math.abs(maxDrawdown) : 0

  return {
    startDate,
    endDate,
    days: base.days,
    years,
    totalReturn,
    cagr: base.cagr,
    vol,
    maxDrawdown,
    calmar,
    sharpe: base.sharpe,
    sortino,
    treynor,
    beta,
    winRate,
    bestDay,
    worstDay,
    avgTurnover,
    avgHoldings,
  }
}

File: rust-indicators/src/backtest/metrics.rs
// src/backtest/metrics.rs
// Performance metrics calculation

use crate::backtest::types::BacktestMetrics;

/// Calculate all backtest metrics from equity curve and returns
pub fn calculate_metrics(
    equity: &[f64],
    daily_returns: &[f64],
    benchmark_returns: &[f64],
    dates: &[String],
    trading_days_per_year: f64,
) -> BacktestMetrics {
    if equity.is_empty() || daily_returns.is_empty() {
        return BacktestMetrics::default();
    }

    let start_date = dates.first().cloned().unwrap_or_default();
    let end_date = dates.last().cloned().unwrap_or_default();
    let days = daily_returns.len() as u32;
    let years = days as f64 / trading_days_per_year;

    // Total return
    let final_equity = *equity.last().unwrap_or(&1.0);
    let total_return = final_equity - 1.0;

    // CAGR
    let cagr = if years > 0.0 {
        final_equity.powf(1.0 / years) - 1.0
    } else {
        0.0
    };

    // Volatility (annualized)
    let vol = annualized_volatility(daily_returns, trading_days_per_year);

    // Max drawdown
    let max_drawdown = calculate_max_drawdown(equity);

    // Calmar ratio
    let calmar = if max_drawdown != 0.0 {
        cagr / max_drawdown.abs()
    } else {
        0.0
    };

    // Sharpe ratio (assuming risk-free rate of 0)
    let sharpe = if vol > 0.0 { cagr / vol } else { 0.0 };

    // Sortino ratio (using downside deviation)
    let downside_vol = downside_deviation(daily_returns, trading_days_per_year);
    let sortino = if downside_vol > 0.0 {
        cagr / downside_vol
    } else {
        0.0
    };

    // Beta and Treynor (relative to benchmark)
    let (beta, treynor) = if !benchmark_returns.is_empty() {
        let b = calculate_beta(daily_returns, benchmark_returns);
        let t = if b != 0.0 { cagr / b } else { 0.0 };
        (b, t)
    } else {
        (1.0, cagr) // Default beta = 1
    };

    // Win rate
    let win_days = daily_returns.iter().filter(|&&r| r > 0.0).count();
    let win_rate = win_days as f64 / daily_returns.len().max(1) as f64;

    // Best/worst day
    let best_day = daily_returns
        .iter()
        .copied()
        .fold(f64::NEG_INFINITY, f64::max);
    let worst_day = daily_returns
        .iter()
        .copied()
        .fold(f64::INFINITY, f64::min);

    BacktestMetrics {
        start_date,
        end_date,
        days,
        years,
        total_return,
        cagr,
        volatility: vol,
        max_drawdown,
        calmar,
        sharpe,
        sortino,
        treynor,
        beta,
        win_rate,
        best_day: if best_day.is_finite() { best_day } else { 0.0 },
        worst_day: if worst_day.is_finite() { worst_day } else { 0.0 },
        avg_turnover: 0.0, // Set separately
        avg_holdings: 0.0, // Set separately
    }
}

/// Calculate annualized volatility
fn annualized_volatility(returns: &[f64], trading_days: f64) -> f64 {
    if returns.is_empty() {
        return 0.0;
    }

    let mean = returns.iter().sum::<f64>() / returns.len() as f64;
    let variance = returns
        .iter()
        .map(|r| (r - mean).powi(2))
        .sum::<f64>()
        / returns.len().max(1) as f64;

    (variance.sqrt()) * trading_days.sqrt()
}

/// Calculate downside deviation (for Sortino ratio)
fn downside_deviation(returns: &[f64], trading_days: f64) -> f64 {
    let negative_returns: Vec<f64> = returns
        .iter()
        .filter(|&&r| r < 0.0)
        .copied()
        .collect();

    if negative_returns.is_empty() {
        return 0.0;
    }

    let sum_sq: f64 = negative_returns.iter().map(|r| r.powi(2)).sum();
    let downside_var = sum_sq / returns.len() as f64;

    downside_var.sqrt() * trading_days.sqrt()
}

/// Calculate maximum drawdown
fn calculate_max_drawdown(equity: &[f64]) -> f64 {
    if equity.is_empty() {
        return 0.0;
    }

    let mut peak = equity[0];
    let mut max_dd = 0.0;

    for &value in equity {
        if value > peak {
            peak = value;
        }
        let dd = (peak - value) / peak;
        if dd > max_dd {
            max_dd = dd;
        }
    }

    -max_dd // Return as negative
}

/// Calculate beta relative to benchmark
fn calculate_beta(returns: &[f64], benchmark: &[f64]) -> f64 {
    if returns.len() != benchmark.len() || returns.is_empty() {
        return 1.0;
    }

    let n = returns.len() as f64;

    let mean_r = returns.iter().sum::<f64>() / n;
    let mean_b = benchmark.iter().sum::<f64>() / n;

    let mut covariance = 0.0;
    let mut variance_b = 0.0;

    for i in 0..returns.len() {
        let r_diff = returns[i] - mean_r;
        let b_diff = benchmark[i] - mean_b;
        covariance += r_diff * b_diff;
        variance_b += b_diff * b_diff;
    }

    covariance /= n;
    variance_b /= n;

    if variance_b > 0.0 {
        covariance / variance_b
    } else {
        1.0
    }
}

/// Calculate turnover between two allocations
pub fn calculate_turnover(
    prev: &std::collections::HashMap<String, f64>,
    curr: &std::collections::HashMap<String, f64>,
) -> f64 {
    let mut total_change = 0.0;

    // Get all tickers from both allocations
    let mut all_tickers: Vec<&String> = prev.keys().chain(curr.keys()).collect();
    all_tickers.sort();
    all_tickers.dedup();

    for ticker in all_tickers {
        let prev_weight = prev.get(ticker).copied().unwrap_or(0.0);
        let curr_weight = curr.get(ticker).copied().unwrap_or(0.0);
        total_change += (curr_weight - prev_weight).abs();
    }

    // Turnover is half the total change (buying and selling are both counted)
    total_change / 2.0
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_max_drawdown() {
        let equity = vec![1.0, 1.1, 1.2, 1.0, 0.8, 1.0, 1.1];
        let dd = calculate_max_drawdown(&equity);
        // Max DD is from 1.2 to 0.8 = -33.3%
        assert!((dd - (-0.333)).abs() < 0.01);
    }

    #[test]
    fn test_annualized_vol() {
        let returns = vec![0.01, -0.01, 0.02, -0.02, 0.01];
        let vol = annualized_volatility(&returns, 252.0);
        assert!(vol > 0.0);
    }

    #[test]
    fn test_beta() {
        let returns = vec![0.01, -0.01, 0.02, -0.02, 0.01];
        let benchmark = vec![0.01, -0.01, 0.02, -0.02, 0.01];
        let beta = calculate_beta(&returns, &benchmark);
        assert!((beta - 1.0).abs() < 0.001); // Same returns = beta 1
    }
}
